File: tictactoe.rb
class Board
  @@winners = [
    [0,1,2],
    [0,4,8],
    [0,3,6],
    [1,4,7],
    [2,5,8],
    [3,4,5],
    [6,7,8],
    [6,4,2]
  ]

  def initialize(board_list=Array.new(9))
    return Error if board_list.class != Array || board_list.length != 9

    @board = board_list
    @x_wins = 0
    @o_wins = 0
    @turns = ["x","o"]
    @turn = 0

    check_winner
  end

  def match(a,b,c)
    a == b && b == c && a != nil
  end

  def check_winner
    @@winners.each do |winning_condition|
      if match @board[winning_condition[0]], @board[winning_condition[1]], @board[winning_condition[2]]

        match @board[winning_condition[0]], @board[winning_condition[1]], @board[winning_condition[2]]

        if @board[winning_condition[0]] == "x"
          @x_wins += 1
        else
          @o_wins += 1
        end
      end
    end

    if @x_wins == 0 && @o_wins == 0
      "no winner"
    elsif @x_wins > @o_wins
      "x wins"
    elsif @x_wins < @o_wins
      "o wins"
    else
      "tie"
    end
  end

  def wipe_board!
    @turn = 0
    @board = Array.new(9)
  end

  def get_board
    puts "#{@board[0] || '_'} #{@board[1] || '_'} #{@board[2] || '_'}"
    puts "#{@board[3] || '_'} #{@board[4] || '_'} #{@board[5] || '_'}"
    puts "#{@board[6] || '_'} #{@board[7] || '_'} #{@board[8] || '_'}"
    self.check_winner
  end

  def set_position char, pos
    raise "You must provide either an 'x' or an 'o'" if !['x','o'].include?(char)

    if @board[pos] == nil
      @board[pos] = char
    else
      raise "That spot is taken!"
    end
  end

  def whose_turn?
    @turns[@turn]
  end

  def take_turn(position)
    self.set_position(@turns[@turn],position)
    @turn = (@turn+1)%2
    get_board
  end
end

File: tictactoe.scala
class Board {
  val winners = List(
    List(0,1,2),
    List(0,4,8),
    List(0,3,6),
    List(1,4,7),
    List(2,5,8),
    List(3,4,5),
    List(6,7,8),
    List(6,4,2)
  )

  val clean_board = Array("_","_","_","_","_","_","_","_","_")

  var board = clean_board
  var x_wins = 0
  var o_wins = 0
  val turns = List("x","o")
  var turn = 0

  check_winner()

  def wipe_board {
    turn = 0
    board = clean_board
  }

  def compare_spots(a: String,b: String,c: String) : Boolean = {
    return a == b && b == c && a != "_"
  }

  def get_board {
    println("")
    println(board(0) + " " + board(1) + " " + board(2))
    println(board(3) + " " + board(4) + " " + board(5))
    println(board(6) + " " + board(7) + " " + board(8))
    println("")
  }

  def check_winner () : String = {
    winners.foreach { winning_condition =>
      if (compare_spots(board(winning_condition(0)),
        board(winning_condition(1)),
        board(winning_condition(2)))) {

        if (board(winning_condition(0)) == "x") {
          x_wins += 1
        } else if (board(winning_condition(0)) == "o") {
          o_wins += 1
        }
      }
    }

    if (x_wins == 0 && o_wins == 0) {
      return "no winner"
    } else if (x_wins > o_wins) {
      return "x wins"
    } else if (x_wins < o_wins) {
      return "o wins"
    } else {
      return "tie"
    }
  }

  def whose_turn : String = {
    return turns(turn)
  }

  def take_turn(position: Int) {
    set_position(turns(turn),position)
    turn = (turn+1)%2
    get_board
    println(check_winner)
  }

  def set_position(char: String, pos: Int) {
    if (board(pos) == "_") {
      board(pos) = char
    } else {
      println("That spot is taken")
    }
  }
}

val board = new Board()

println(board.check_winner)

board.take_turn(1)
board.take_turn(2)
board.take_turn(3)

board.get_board